[ENH] Add `_ConditionalRefitForecaster` and refactor `UpdateRefitsEvery` to subclass it

ZiyaoWei opened the pull request:
#### Reference Issues/PRs

Initial steps of #652.

#### What does this implement/fix? Explain your changes.

This adds a `_ConditionalRefitForecaster` as mentioned [here](https://github.com/alan-turing-institute/sktime/issues/652#issuecomment-772585593), to be used in implementations of different refitting forecasters.

#### Does your contribution introduce a new dependency? If yes, which one?

No.

#### What should a reviewer concentrate their feedback on?

The API design - is this okay and can it be exposed to the public?

#### PR checklist

##### For all contributions
- [ ] I've added myself to the [list of contributors](https://github.com/alan-turing-institute/sktime/blob/main/.all-contributorsrc).
- [ ] Optionally, I've updated sktime's [CODEOWNERS](https://github.com/alan-turing-institute/sktime/blob/main/CODEOWNERS) to receive notifications about future changes to these files.
- [ ] I've added unit tests and made sure they pass locally.
- [ ] The PR title starts with either [ENH], [MNT], [DOC], or [BUG] indicating whether the PR topic is related to enhancement, maintenance, documentation, or bug.

##### For new estimators
- [ ] I've added the estimator to the online documentation.
- [ ] I've updated the existing example notebooks or provided a new one to showcase how my estimator works.

fkiraly commented:
Nice! May I suggest:

* make the `ConditionalRefitForecaster` a separate estimator
* don't start with an underscore, or tests framework will not detect it
* leave `UpdateRefitsEvery` unchanged per functionality, as we have already released it. Fine if you refactor it (subclass or common superclass) but please ensure that the functionality does not change, at least for the next deprecation cycle.

ZiyaoWei commented:
Thanks for the quick response! To make sure I understand:

> * make the `ConditionalRefitForecaster` a separate estimator
Do you mean putting it in a separate file?

> * don't start concrete estimators with an underscore, or tests framework will not detect it. Joint superclasses can be private, of course.
So we are making `ConditionalRefitForecaster` public, and users would subclass it? (see [this comment](https://github.com/alan-turing-institute/sktime/issues/652#issuecomment-1200247799) as well)

> * leave `UpdateRefitsEvery` unchanged per functionality, as we have already released it. Fine if you refactor it (subclass or common superclass) but please ensure that the functionality does not change, at least for the next deprecation cycle.
I'd love to add some tests for it to make sure the functionality doesn't change. Maybe I can file a ticket for adding tests and refactoring?

fkiraly commented:
> Do you mean putting it in a separate file?

No, just a separate class.

> So we are making `ConditionalRefitForecaster` public, and users would subclass it? (see [this comment](https://github.com/alan-turing-institute/sktime/issues/652#issuecomment-1200247799) as well)

answering there

> I'd love to add some tests for it to make sure the functionality doesn't change. Maybe I can file a ticket for adding tests and refactoring?

good idea. Please note that generic tests already cover the forecaster, so the only things that need to be tested are whether the functionality is correct.

ZiyaoWei commented:
Thanks @fkiraly! I've updated the PR to remove the refactoring and added a public class `ConditionalRefitForecaster`, which just takes in a bunch of functions and pass them into `_ConditionalRefitForecaster`. Hope this makes some sense and further discussions are most welcome!

@fkiraly - you mention `although there may be a case for or against having a joint parent class` in the comments in #652 - can you elaborate a bit? There are currently both the parent class and the public class in the PR, but I'll be happy to remove one of the classes if it proves to be not of much use.

fkiraly commented:
> @fkiraly - you mention although there may be a case for or against having a joint parent class in the comments in https://github.com/alan-turing-institute/sktime/issues/652 - can you elaborate a bit? There are currently both the parent class and the public class in the PR, but I'll be happy to remove one of the classes if it proves to be not of much use.

I thought the parent could be a parent of `UpdateRefitsEvery` as well, or be disconnected of it. Not sure what is better.

ZiyaoWei commented:
I'll change this to ready for review - I'll add tests for this but want to make sure it looks reasonable at least.

fkiraly commented:
Hm, overall looks nice!

I think though the logic of the condition and data retrieval can be simplified, could be wrapped in less functions.

Is it not possible to achieve the same effect by a single function that checks the condition, for the simplest version?
And perhaps second function for the data memory or subsetting.

Could you explain what the `_set_refit_condition` and `_update_refit_condition` are meant to do?
